### PR TITLE
Support uap10.0.16299

### PR DESCRIPTION
**Pull request by StephenBonikowsky**
Trying to execute on option 2 as described in https://github.com/dotnet/wcf/issues/2687

Referencing https://github.com/dotnet/corefx/pull/27787 but many of the changes do not directly apply.

Also please note that the buildtools version being used was forked several months ago so it is possible/likely some issues will arise because of that.
In this current state I am getting...

```
C:\OSS\wcf\build.proj(42,5): error MSB4018: The "GenerateConfigurationProps" task failed unexpectedly.\r
C:\OSS\wcf\build.proj(42,5): error MSB4018: System.Exception: Unknown compatible value >uap10.0.16299 for property TargetGroup value uap10.0.16299aot.\r
C:\OSS\wcf\build.proj(42,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.PropertyValue.ConnectValues(String property, Dictionary`2 values)\r
C:\OSS\wcf\build.proj(42,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.PropertyValue.ConnectValues(Dictionary`2 values)\r
C:\OSS\wcf\build.proj(42,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.ConfigurationFactory..ctor(ITaskItem[] properties, ITaskItem[] propertyValues)\r
C:\OSS\wcf\build.proj(42,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.ConfigurationTask.LoadConfiguration()\r
C:\OSS\wcf\build.proj(42,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.GenerateConfigurationProps.Execute()\r
C:\OSS\wcf\build.proj(42,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()\r
C:\OSS\wcf\build.proj(42,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

@ericstj and @joperezr I could use your feedback.

### CODE REVIEW

**Comment by StephenBonikowsky** [comment]
I've made sure my local environment is completely clean after making changes and before re-running.

File: external/uap/uap.depproj [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    <NuGetDeploySourceItem>Reference</NuGetDeploySourceItem>
+    <UAPPackageVersion>6.0.7</UAPPackageVersion>
+    <TargetFramework>uap10.0.16299</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NETCore.UniversalWindowsPlatform">
+      <Version>$(UAPPackageVersion)</Version>
+    </PackageReference>
+    <!-- System.Reflection.Emit and System.Reflection.Emit.ILGeneration are needed for the closure of the implementation of System.Reflection.DispatchProxy.
```
Comment (joperezr) [inline external/uap/uap.depproj]: I don't think you need Ref.Emit nor ILGeneration for your projects, so you should be able to remove everything from lines 17-28.

**Comment by joperezr** [comment]
> In this current state I am getting.

I didn't check on the review but I believe that the problem you are getting might be related to having a TAB instead of spaces on one of your new configurations on your configurations.props. Check and make sure that they are all spaces since our logic will only trim out spaces and not tabs.

File: external/dir.proj [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
   <!-- Build for all configurations -->
   <ItemGroup>
+    <Project Condition="'$(BuildAllConfigurations)' == 'true'" Include="uap/uap.depproj" />
```
[the remark this answers is not in the document]
Comment (StephenBonikowsky) [reply external/dir.proj]: In VSO our build command is...
```
build.cmd -OfficialBuildId=20180324-01 -allconfigurations -buildArch=x64 -Debug -- /p:SignType=real /p:RuntimeOS=win10 /p:EnableProfileGuidedOptimization=true
```

And it produces the packages.

File: external/runtime/runtime.depproj [hunk 3]
```diff
@@ -10,12 +10,12 @@
     <RidSpecificAssets>true</RidSpecificAssets>
   </PropertyGroup>
 
-  <PropertyGroup Condition="'$(TargetGroup)' == 'uapaot'">
+  <PropertyGroup Condition="'$(TargetGroup)' == 'uapaot' Or '$(TargetGroup)' == 'uap10.0.16299aot'">
```
[the remark this answers is not in the document]
Comment (StephenBonikowsky) [reply external/runtime/runtime.depproj]: Wasn't sure, I'll remove it.

File: src/System.ServiceModel.Duplex/src/Configurations.props [hunk 4]
```diff
@@ -8,6 +8,7 @@
     <BuildConfigurations>
       $(PackageConfigurations);
       uap-Windows_NT;
+      uap10.0.16299-Windows_NT;
```
[the remark this answers is not in the document]
Comment (StephenBonikowsky) [reply src/System.ServiceModel.Duplex/src/Configurations.props]: Can you explain why it should be in the ```<PackageConfigurations>``` property for the 5 facades when ```uap-Windows_NT;``` was not?

And is it correct to keep ```uap-Windows_NT;``` or should I remove it?

**Comment by StephenBonikowsky** [comment]
Getting close, I'm currently getting a compile error because...
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin\Microsoft.Common.CurrentVersion.targets(1964,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "Windows". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\OSS\wcf\src\System.Private.ServiceModel\src\System.Private.ServiceModel.csproj]
```

Looks like a restore problem, haven't yet found what is different from when we just target ```uap```.

**Comment by StephenBonikowsky** [comment]
Now we are at the package validation stuff...

For NetTcp, Http, Duplex and Security I am getting the following validation errors...
`C:\OSS\wcf\Tools\Packaging.targets(1103,5): error : When targeting UAP,Version=v10.0.16300/win10-x86 System.ServiceModel.NetTcp will use C:\OSS\wcf\bin\Windows_NT.AnyCPU.Release\System.ServiceModel.NetTcp\uap10.0.16299\System.ServiceModel.NetTcp.dll which targets uap10.0.16299  but lib/uap10.0.16300/System.ServiceModel.NetTcp.dll targets uap10.0.16300 specifically. [C:\OSS\wcf\src\System.ServiceModel.NetTcp\pkg\System.ServiceModel.NetTcp.pkgproj]`

The same packages are also complaining about the API version it should support as stated in the packageindex.json as compared to the package it finds in under WCF\packages\
`C:\OSS\wcf\Tools\Packaging.targets(1103,5): error : System.ServiceModel.NetTcp should support API version 4.1.3.0 on UAP,Version=v10.0.16299 but C:\OSS\wcf\packages\System.ServiceModel.NetTcp\4.3.0\lib\netcore50\System.ServiceModel.NetTcp.dll was found to support 4.1.1.0. [C:\OSS\wcf\src\System.ServiceModel.NetTcp\pkg\System.ServiceModel.NetTcp.pkgproj]`

**Comment by StephenBonikowsky** [comment]
@ericstj made some changes based on realizing that we don't need to build the 5 contracts any differently for uap.

**Comment by StephenBonikowsky** [comment]
Woot! Thank you for all your help getting this working.

**Pull request merged**